### PR TITLE
Adapt hyperparam example to new MLflow APIs.

### PR DESCRIPTION
**Pull request by tomasatdatabricks**
## What changes are proposed in this pull request?
 Adapted hyperparam example to new mlflow apis.

 
## How is this patch tested?
Only manually, the same as other examples.
### Is this a user-facing change? 

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [x] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/hyperparam/search_gpyopt.py [hunk 1]
```diff
@@ -89,33 +89,32 @@ def eval(params):
...
-                    "epochs": str(nepochs),
-                    "learning_rate": str(lr),
-                    "momentum": str(momentum),
-                    "seed": str(seed)},
-                experiment_id=experiment_id,
-                synchronous=False
-            )
-
-            if p.wait():
+            with mlflow.start_run(nested=True) as child_run:
```
Comment (smurching) [inline examples/hyperparam/search_gpyopt.py]: Hm why is this necessary? It doesn't look like `projects.run` mutates the current active run (i.e.  it calls `create_run` instead of `start_run`, so we should be able to just call `projects.run` I think? Unless there's something I'm missing

File: examples/hyperparam/search_random.py [hunk 2]
```diff
@@ -89,47 +84,37 @@ def get_metric(metric_name):
...
+        best_val_test = _inf
+        best_run = None
+        for r in runs:
+            if r.data.metrics["val_rmse"] < best_val_valid:
+                best_run = r
+                best_val_train = r.data.metrics["train_rmse"]
+                best_val_valid = r.data.metrics["val_rmse"]
+                best_val_test = r.data.metrics["test_rmse"]
+        mlflow.set_tag("best_run", best_run.info.run_id)
+        mlflow.log_metric("train_{}".format(metric), best_val_train)
```
Comment (smurching) [inline examples/hyperparam/search_random.py]: We could optionally use `mlflow.log_metrics({"train_{}".format(metric): best_val_train, "val_{}".format(metric): best_val_valid, "test_{}".format(metric): best_val_test})` here as an example of how to log multiple metrics in a single call, but also fine to leave as is.

**Review by smurching — APPROVED** [review]
LGTM given that this works via manual testing. I do wonder if there's a bug in projects that we should fix - you shouldn't have to start the run for a project outside the call to `projects.run`

File: examples/hyperparam/search_gpyopt.py (hunk 1 above)
Comment (tomasatdatabricks) [reply examples/hyperparam/search_gpyopt.py]: This is necessary so to keep have the parent run id set in the child run. process.run does not do it (I tired to pass in the parent run id but that does not work at all). The alternative would be to use the client api to set the tag manually, but I think this one is cleaner, especially if we would ever decide to inherit more tags.

File: examples/hyperparam/search_random.py (hunk 2 above)
Comment (tomasatdatabricks) [reply examples/hyperparam/search_random.py]: hmm that's a good point. I think we should do it. It's an example, it should show the best of mlflow! ;)
